src/common/utils/reloadLib.ts

```diff
@@ -0,0 +1,18 @@
+import {ethers, CreateXOpts, Deployments} from "@interlay/xopts";
+import {loadLibAction} from "../actions/lib.actions";
+import {Signer} from "@interlay/xopts/dist/lib/core";
+import {Dispatch} from "react";
+
+// eslint-disable-next-line
+const detectEthereumProvider = require("@metamask/detect-provider");
+
+export const fireLoading = async (dispatch: Dispatch<any>, useMock: boolean): Promise<void> => {
+    const web3 = await detectEthereumProvider();
+    window.provider = new ethers.providers.Web3Provider(web3);
+    const isSigner = window.provider instanceof Signer;
+
+    const addresses = useMock ? Deployments.mock : undefined;
+    window.xopts = await CreateXOpts(window.provider, addresses);
```

Comment:
I'm not a big fan of using the `window` object directly.
I think I would rather have a singleton-like module to keep this information.
Probably something along the lines of:

```typescript
class XOptsContainer {
  private _xopts: XOpts<ProviderOrSigner>;
  public provider: Provider;
  
  set xopts(xopts: XOpts<ProviderOrSigner>) {
    this._xopts = xopts;
  }

  get xopts(): XOpts<Provider> {
    return this._xopts;
  }

  get xoptsRW(): XOpts<Signer> {
    // runtime typecheck/cast
    return this._xopts;
  }
}

const xoptsContainer = new XOptsContainer();
export default xoptsContainer;

// usage
import lib from xoptsLib;

lib.xopts.options.list();
```

Reply:
I think this is a fair argument; I can add this easily enough.
Not sure I understand your usage - the way I see it, it'd be similar to the current setup, except you'd do `xoptsContainer.xopts()` rather than `window.xopts`.

Reply:
I think in the current state `window.xopts.options.write(...)` will not work because of the type.
The usage is assuming this is saved in a file called `xoptsLib` but it could be `xoptsContainer`, I just found the name a bit long.
`xopts` is a getter, so `xopts` should be enough, no need for the `()`.
I did not test the code though, so it could be a bit off.